klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+
+    if len(results_df_list) > 0:
+        results_df = pd.concat(results_df_list)
+    else:
+        results_df = pd.DataFrame()
+
+    return results_df, tc_df
+
+
+def parse_existing_rules(rule_deck_path, output_path, target_table=None):
+    """
+    This function collects the rule names from the existing drc rule decks.
+
+    Parameters
+    ----------
+    rule_deck_path : string or Path object
+        Path string to the DRC directory where all the DRC files are located.
+    output_path : string or Path
+        Path of the run location to store the output analysis file.
+    target_table : string Optional
+        Name of the table to be in testing
+
+    Returns
+    -------
+    pd.DataFrame
+        A pandas DataFrame with the rule and rule deck used.
+    """
+
+    if target_table is None:
+        drc_files = glob.glob(os.path.join(rule_deck_path, "rule_decks", "*.drc"))
+    else:
+        table_rule_file = os.path.join(rule_deck_path, "rule_decks", f"{target_table}.drc")
+        if not os.path.isfile(table_rule_file):
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), table_rule_file)
+
+        drc_files = [table_rule_file]
+
+    rules_data = list()
+
+    for runset in drc_files:
+        with open(runset, "r") as f:
+            for line in f:
+                if ".output" in line:
+                    line_list = line.split('"')
+                    rule_info = dict()
+                    rule_info["table_name"] = os.path.basename(runset).replace(".drc", "")
+                    rule_info["rule_name"] = line_list[1]
+                    rule_info["in_rule_deck"] = 1
+                    rules_data.append(rule_info)
+
+    df = pd.DataFrame(rules_data)
+    df.drop_duplicates(inplace=True)
+    df.to_csv(os.path.join(output_path, "rule_deck_rules.csv"), index=False)
+    return df
+
+
+def analyze_test_patterns_coverage(rules_df, tc_df, output_path):
+    """
+    This function analyze the test patterns before running the test cases.
+
+    Parameters
+    ----------
+    rules_df : pd.DataFrame
+        DataFrame that holds all the rules that are found in the rule deck.
+    tc_df : pd.DataFrame
+        DataFrame that holds all the test cases and all the information required.
+    output_path : string or Path
+        Path of the run location to store the output analysis file.
+
+    Returns
+    -------
+    pd.DataFrame
+        A DataFrame with analysis of the rule testing coverage.
+    """
+    cov_df = (
+        tc_df[["table_name", "rule_name"]]
+        .groupby(["table_name", "rule_name"])
+        .count()
+        .reset_index(drop=False)
+    )
+    cov_df = cov_df.merge(rules_df, on="rule_name", how="outer")
+    cov_df["runset"].fillna("", inplace=True)
+    cov_df.to_csv(os.path.join(output_path, "testcases_coverage.csv"), index=False)
+
+    return cov_df
+
+
+def generate_merged_testcase(orignal_testcase, marker_testcase):
```

Comment:
curious if the same could be achieved with `klayout.db`

Reply:
As I mentioned in another comment, gdstk is much faster than klayout and handles larger designs with much smaller footprint. Klayout has a very large memory footprint and much slower in handling data. I only choose klayout if the required task is not easily supported by gdstk.

Reply:
#11